File: src/tools/explain.ts
import { z } from "zod";
import type { RequestHandlerExtra } from "@modelcontextprotocol/sdk/shared/protocol.js";
import type {
  ServerRequest,
  ServerNotification,
} from "@modelcontextprotocol/sdk/types.js";

import { BackendIdSchema } from "../types.js";
import { handleRescue, type RescueDeps } from "./rescue.js";

export const explainInputShape = {
  target: z
    .string()
    .min(1)
    .describe("Target to explain: a path, glob, function name, or topic."),
  depth: z
    .enum(["brief", "deep"])
    .optional()
    .describe("Explanation depth (default: brief)."),
  backend: BackendIdSchema.optional(),
  cwd: z.string().optional(),
};

export type ExplainArgs = {
  target: string;
  depth?: "brief" | "deep";
  backend?: z.infer<typeof BackendIdSchema>;
  cwd?: string;
};

export const handleExplain = async (
  args: ExplainArgs,
  extra: RequestHandlerExtra<ServerRequest, ServerNotification>,
  deps: Omit<RescueDeps, "routeKey">,
): ReturnType<typeof handleRescue> => {
  const depth = args.depth ?? "brief";
  const lines = [
    `Explain: ${args.target}`,
    depth === "brief"
      ? "Keep the explanation tight (under 200 words). Lead with purpose, then mechanism."
      : "Provide a deep walk-through: design rationale, control flow, edge cases, and how it interacts with the surrounding system.",
  ];
  return handleRescue(
    {
      task: lines.join("\n"),
      backend: args.backend,
      cwd: args.cwd,
    },
    extra,
    { ...deps, routeKey: "explain" },
  );
};

File: src/cli/explain.rs
use anyhow::Result;

use crate::config::RouteKey;
use crate::types::BackendId;

pub async fn run(
    target: String,
    deep: bool,
    backend: Option<BackendId>,
    cwd: Option<String>,
) -> Result<()> {
    let mut lines = vec![
        "You are working in a codebase rooted at the current working directory.".to_string(),
        String::new(),
        format!("Explain: {target}"),
        String::new(),
        "Workflow — MUST follow. Do not produce an explanation without reading.".to_string(),
        "1. Locate and read <target> in full (a file, symbol, module, or concept).".to_string(),
        "2. Read related files (callers, callees, neighbouring modules) needed to explain it accurately. Speculation without reading the surrounding code is not acceptable.".to_string(),
        "3. Cite file:line when it helps the reader navigate.".to_string(),
        String::new(),
    ];
    lines.push(if deep {
        "Provide a deep walk-through: design rationale, control flow, edge cases, and how it interacts with the surrounding system.".to_string()
    } else {
        "Keep the explanation tight (under 200 words). Lead with purpose, then mechanism.".to_string()
    });
    super::rescue::run_with_route(lines.join("\n"), backend, cwd, true, RouteKey::Explain).await
}
